File: 09_07_2024_Closest_Three_Sum.py

class Solution:
    def threeSumClosest(self, arr, target):
        arr.sort()
        n = len(arr)
        ans = float('inf')
        diff = float('inf')

        for i in range(n - 2):
            st, end = i + 1, n - 1
            while st < end:
                sum_ = arr[i] + arr[st] + arr[end]
                current_diff = abs(sum_ - target)
                if current_diff < diff or (current_diff == diff and sum_ > ans):
                    diff = current_diff
                    ans = sum_
                if sum_ < target:
                    st += 1
                elif sum_ > target:
                    end -= 1
                else:
                    return target
        return ans

File: 09_07_2024_Closest_Three_Sum.java
class Solution {
    static int threeSumClosest(int[] arr, int target) {
        // code here
        int diff = Integer.MAX_VALUE;
        int closestSum = Integer.MAX_VALUE;

        Arrays.sort(arr);

        for (int i = 0; i < arr.length; i++) {
            int s = i + 1, e = arr.length - 1;
            while (s < e) {
                int sum = arr[i] + arr[s] + arr[e];
                if (sum == target) {
                    return sum;
                } else if (sum > target) {
                    e--;
                } else {
                    s++;
                }

                if (Math.abs(sum - target) < diff) {
                    diff = Math.abs(sum - target);
                    closestSum = sum;
                } else if (Math.abs(sum - target) == diff) {
                    closestSum = Math.max(closestSum, sum);
                }
            }
        }

        return closestSum;
    }
}
